File: WQChartsDemo/Objc/View/ListCell.m
// 代码地址: 
// ListCell.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "ListCell.h"

@interface ListCell()

@property (nonatomic) CGFloat buttonHeight;

@property (nonatomic, strong) ListView* contentView;
@property (nonatomic, strong) UILabel* titleLabel;
@property (nonatomic, strong) UIButton* appendButton;
@property (nonatomic, strong) UIButton* removeButton;

@end

@implementation ListCell

#pragma mark - Public

- (instancetype)initWithFrame:(CGRect)frame {
    if(self = [super initWithFrame:frame]) {
        self.buttonHeight = 25;
        self.backgroundColor = UIColor.clearColor;
        
        [self addSubview:self.titleLabel];
        [self addSubview:self.contentView];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat width = self.bounds.size.width;
    CGFloat height = self.bounds.size.height;
    
    CGFloat insetTop = 10;
    
    UILabel* textLabel = self.titleLabel;
    [textLabel sizeToFit];
    textLabel.frame = CGRectMake(10, insetTop, width - 20, textLabel.bounds.size.height);
    insetTop = CGRectGetMaxY(textLabel.frame);
    
    CGFloat insetBottom = 10;
    if (self.isMutable) {
        CGFloat buttonWidth = width / 2;
        CGFloat buttonHeight = self.buttonHeight;
        _appendButton.frame = CGRectMake(0, height - insetBottom - buttonHeight, buttonWidth, buttonHeight);
        _removeButton.frame = CGRectMake(buttonWidth, height - insetBottom - buttonHeight, buttonWidth, buttonHeight);
        insetBottom += buttonHeight;
    }
    
    self.contentView.frame = CGRectMake(0, insetTop, width, height - insetTop - insetBottom);
}


- (CGSize)sizeThatFits:(CGSize)size {
    CGFloat height = 10;
    [self.titleLabel sizeToFit];
    height += self.titleLabel.bounds.size.height;
    height += [self.contentView sizeThatFits:CGSizeMake(size.width, CGFLOAT_MAX)].height;
    if(self.isMutable) {
        height += self.buttonHeight;
    }
    height += 10;
    return CGSizeMake(size.width, MIN(size.height, height));
}


- (ListCell*(^)(NSString*))setTitle {
    return ^ListCell *(NSString *title) {
        self.title = title;
        return self;
    };
}

- (ListCell*(^)(BOOL))setIsMutable {
    return ^ListCell *(BOOL isMutable) {
        self.isMutable = isMutable;
        return self;
    };
}

- (ListCell*(^)(NSArray<UIView*>*))setItems {
    return ^ListCell *(NSArray<UIView*>* subcells) {
        self.contentView.items = subcells;
        return self;
    };
}

- (ListCell*(^)(void(^)(ListCell*)))setOnAppend {
    return ^ListCell*(void(^onAppend)(ListCell*)) {
        self.onAppend = onAppend;
        return self;
    };
}

- (ListCell*(^)(void(^)(ListCell*)))setOnRemove {
    return ^ListCell*(void(^onRemove)(ListCell*)) {
        self.onRemove = onRemove;
        return self;
    };
}

- (ListCell*(^)(UIView*))addItem {
    return ^ListCell *(UIView* subcell) {
        [self.contentView addItem:subcell];
        return self;
    };
}

- (ListCell*(^)(NSInteger))removeItemAtIndex {
    return ^ListCell *(NSInteger index) {
        [self.contentView removeItemAt:index];
        return self;
    };
}

#pragma mark - Getter


- (UILabel *)titleLabel {
    if(!_titleLabel) {
        _titleLabel = UILabel.new;
        _titleLabel.font = [UIFont systemFontOfSize:10];
        _titleLabel.textColor = Color_White;
    }
    return _titleLabel;
}

- (ListView *)contentView {
    if(!_contentView) {
        _contentView = ListView.new;
    }
    return _contentView;
}


#pragma mark - Setter

- (void)setTitle:(NSString *)title {
    _title = title;
    self.titleLabel.text = title;
    [self setNeedsLayout];
}

- (void)setIsMutable:(BOOL)isMutable {
    _isMutable = isMutable;
    
    if (isMutable) {
        if (!_appendButton) {
            UIButton* appendButton = [UIButton buttonWithType:UIButtonTypeCustom];
            appendButton.backgroundColor = Color_Tint;
            appendButton.titleLabel.font = [UIFont systemFontOfSize:9];
            [appendButton setTitle:@"Append" forState:UIControlStateNormal];
            [appendButton addTarget:self action:@selector(onAppendButtonClick) forControlEvents:UIControlEventTouchUpInside];
            [self addSubview:appendButton];
            _appendButton = appendButton;
        }
        
        if (!_removeButton) {
            UIButton* removeButton = [UIButton buttonWithType:UIButtonTypeCustom];
            removeButton.backgroundColor = Color_Red;
            removeButton.titleLabel.font = [UIFont systemFontOfSize:9];
            [removeButton setTitle:@"Remove" forState:UIControlStateNormal];
            [removeButton addTarget:self action:@selector(onRemoveButtonClick) forControlEvents:UIControlEventTouchUpInside];
            [self addSubview:removeButton];
            _removeButton = removeButton;
        }
    } else {
        [_appendButton removeFromSuperview];
        _appendButton = nil;
        
        [_removeButton removeFromSuperview];
        _removeButton = nil;
    }
    
    [self setNeedsLayout];
}

#pragma mark - Private

- (void)onAppendButtonClick {
    if(self.onAppend) {
        self.onAppend(self);
    }
}

- (void)onRemoveButtonClick {
    if(self.onRemove) {
        self.onRemove(self);
    }
}

@end

File: WQChartsDemo/Swift/View/ListCell.swift
// 代码地址: https://github.com/CoderWQYao/WQCharts-iOS
//
// ListCell.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class ListCell: UIView {
    
    open lazy var titleLabel: UILabel = {
        let titleLabel = UILabel()
        titleLabel.font = UIFont.systemFont(ofSize: 10)
        titleLabel.textColor = Color_White
        return titleLabel
    }()
    
    open lazy var contentView: ListView = {
        let contentView = ListView()
        return contentView
    }()
    
    var _appendButton: UIButton?
    var _removeButton: UIButton?
    var buttonHeight = CGFloat(25)
    
    open var title: String? {
        didSet {
            titleLabel.text = title
            setNeedsLayout()
        }
    }
    
    open func setTitle(_ title: String?) -> ListCell {
        self.title = title
        return self
    }
    
    open var isMutable = false {
        didSet {
            if isMutable {
                if _appendButton == nil {
                    let appendButton = UIButton()
                    appendButton.backgroundColor = Color_Tint
                    appendButton.titleLabel?.font = UIFont.systemFont(ofSize: 9)
                    appendButton.setTitle("Append", for: .normal)
                    appendButton.addTarget(self, action: #selector(onAppendButtonClick), for: .touchUpInside)
                    addSubview(appendButton)
                    _appendButton = appendButton
                }
                
                if _removeButton == nil {
                    let removeButton = UIButton()
                    removeButton.backgroundColor = Color_Red
                    removeButton.titleLabel?.font = UIFont.systemFont(ofSize: 9)
                    removeButton.addTarget(self, action: #selector(onRemoveButtonClick), for: .touchUpInside)
                    removeButton.setTitle("Remove", for: .normal)
                    addSubview(removeButton)
                    _removeButton = removeButton
                }
            } else {
                if let addButton = _appendButton {
                    addButton.removeFromSuperview()
                    _appendButton = nil
                }
                if let removeButton = _removeButton {
                    removeButton.removeFromSuperview()
                    _removeButton = nil
                }
            }
            setNeedsLayout()
        }
    }
    
    open func setIsMutable(_ isMutable: Bool) -> ListCell {
        self.isMutable = isMutable
        return self
    }
    
    var onAppend: ((_ sliderListView: ListCell) -> Void)?
    open func setOnAppend(_ onAppend: ((_ sliderListView: ListCell) -> Void)?) -> ListCell {
        self.onAppend = onAppend
        return self
    }
    
    var onRemove: ((_ sliderListView: ListCell) -> Void)?
    open func setOnRemove(_ onRemove: ((_ sliderListView: ListCell) -> Void)?) -> ListCell {
        self.onRemove = onRemove
        return self
    }
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        backgroundColor = .clear
        
        addSubview(titleLabel)
        addSubview(contentView)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        
        let width = bounds.width
        let height = bounds.height

        var insetTop: CGFloat = 10
        titleLabel.sizeToFit()
        titleLabel.frame = CGRect(x: 10, y: insetTop, width: width - 20, height: titleLabel.bounds.height)
        insetTop = titleLabel.frame.maxY

        var insetBottom: CGFloat = 10
        if isMutable {
           let buttonWidth = width / 2
           let buttonHeight = self.buttonHeight
           _appendButton?.frame = CGRect(x: 0, y: height - insetBottom - buttonHeight, width: buttonWidth, height: buttonHeight)
           _removeButton?.frame = CGRect(x: buttonWidth, y: height - insetBottom - buttonHeight, width: buttonWidth, height: buttonHeight)
           insetBottom += buttonHeight
        }

        contentView.frame = CGRect(x: 0, y: insetTop, width: width, height: height - insetTop - insetBottom)
    }
  
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        var height: CGFloat = 10
        titleLabel.sizeToFit()
        height += titleLabel.bounds.height
        height += self.contentView.sizeThatFits(CGSize(width: size.width, height: CGFloat.greatestFiniteMagnitude)).height
        if isMutable {
            height += buttonHeight
        }
        height += 10
        return CGSize(width: size.width, height: min(size.height, height))
    }
    
    
    open func addItem(_ item: UIView) -> ListCell{
        contentView.addItem(item)
        return self
    }
    
    open func removeItem(at index: Int) -> ListCell {
        contentView.removeItem(at: index)
        return self
    }
    
    @objc func onAppendButtonClick() {
        guard let onAppend = onAppend else {
            return
        }
        onAppend(self)
    }
    
    @objc func onRemoveButtonClick() {
        guard let onRemove = onRemove else {
            return
        }
        onRemove(self)
    }
    
}
